### ISSUE TITLE
Use File Attachment annotation for the link to the embedded file

### ISSUE THREAD

**Issue by Intelligent2013** [metanorma/mn2pdf#267]
Source issue: https://github.com/metanorma/metanorma/issues/407#issuecomment-2321962832

> For TC4, the type of PDF annotation generated is incorrect.
> 
> It is currently a Link annotation (`/Subtype /Link` supporting a URI via the `/A` (Action) key) whereas it should likely be a File Attachment annotation (`/Subtype /FileAttachment` with `/FS` entry to an embedded file object). Although technically an Embedded GoTo action is possible using the current Link annotation (by changing the `/A`' entry), I'm not sure how widely supported that feature is in browser-based PDF viewers and the less widely used PDF viewers.

This is how the Apache FOP is working (the source code https://github.com/apache/xmlgraphics-fop/blob/c11f43c9dbf529b87820d77ef501aa10699fc9d8/fop-core/src/main/java/org/apache/fop/pdf/PDFFactory.java#L699)
Apache FOP add JavaScript action to open the embedded file:

PDFBox PDFDebugger:
![image](https://github.com/user-attachments/assets/e96b68d7-4a58-4b37-80b7-735c7e3f467a)

Adobe Acrobat:
![image](https://github.com/user-attachments/assets/10c01dc1-ba54-4bd2-a919-698ea6f23f1d)

From PDF Reference 1.7:
![image](https://github.com/user-attachments/assets/69cf935a-aa93-4eeb-bdbf-59ba927c974c)

![image](https://github.com/user-attachments/assets/62374082-18f6-4c03-827b-3cba0e3cd647)

To do:
- [x] change  `/Subtype /Link` to the File Attachment annotation (`/Subtype /FileAttachment` with `/FS` entry to an embedded file object

**Comment by Intelligent2013** [comment]
The `FileAttachment` annotation is working differently than text with hyperlink with Javascript function (current):
- the active area is icon (Graph, Paperclip, PushPin or Tag) vs. text hyperlink
- the attachment opens after double-click on the icon vs. just click on the text

Example from `ISO_32000-2_sponsored-ec2.pdf` with two 'paper clips' links:
![image](https://github.com/user-attachments/assets/989b1d04-6c46-4444-80ab-60a9cb4061a8)


Another very simple PDF with `/Subtype /FileAttachment`: 
[structure-attached.pdf](https://github.com/user-attachments/files/16998792/structure-attached.pdf)
(found on GH somewhere, opening in the Adobe Acrobat only, not well-formed PDF)

Additional tasks:
- [ ] remove blue color and action from the link
- [ ] add icon after the link.

**Comment by Intelligent2013** [comment]
Current progress, proof-of-concept example:  [test_attachments.tc4.presentation.pdf](https://github.com/user-attachments/files/17007511/test_attachments.tc4.presentation.pdf)

1. the Adobe Acrobat shows the message:
![image](https://github.com/user-attachments/assets/cd400857-8e91-4dc6-ab66-24d9a7bb4534)

The double-click on the 'Paperclip' icon doesn't work until the user click on the button 'Enable editing'.

2.  The 'Attachment' tab contains two files - attached file and 2nd one shows when annotation is created. I'll investigate how to omit the second one.

![image](https://github.com/user-attachments/assets/2ae2c98f-6d16-41e1-8db4-a787440f2162)



Note: We can't generate PDF/A-3 for BIPM brochure, because  the link to the embedded .mml file realized in the Apache FOP via JavaScript function, that doesn't allow by PDF/A-3.

**Comment by petervwyatt** [comment]
That last file is a special file from Ange (a file format expert who experiments with minimal files)...

Other PDF files with file attachments on the public web:
* https://bugs.ghostscript.com/attachment.cgi?id=7736
* [Isartor Test Suite for PDF/A](https://pdfa.org/isartor-test-suite-terms-of-use-download/) - isartor-6-5-2-t02-fail-a.pdf (it's a valid PDF but a fail case 'cos earlier PDF/A versions did not allow file annotations)
* various PDFs from https://github.com/openpreserve/format-corpus/tree/master/pdfCabinetOfHorrors such as [this one](https://github.com/openpreserve/format-corpus/blob/master/pdfCabinetOfHorrors/fileAttachment.doc)
* the veraPDF corpus (both pass and fail examples, depending on conformance level) - e.g. https://github.com/veraPDF/veraPDF-corpus/tree/staging/PDF_A-3b/6.8%20Embedded%20files

**Comment by petervwyatt** [comment]
@Intelligent2013 
> The double-click on the 'Paperclip' icon doesn't work until the user click on the button 'Enable editing'.

Correct. That is a PDF/A formal requirement to ensure that any user interaction (such as extracting an embedded file) cannot / does not mess with page content. This is not a bug :-)

**Comment by Intelligent2013** [comment]
@petervwyatt thank you for the clarification and the examples.

**Comment by Intelligent2013** [comment]
So, how the PDF viewer (Adobe Acrobat, Foxit Reader) shows the attached files:
- each `FileAttachment` annotation has the entry in the `Attachment` tab, i.e.
    - if two links on the different pages point to the same attachment, then there are two identical entries (except page number)
    - if two links on the one page point to the same attachment, then there are two identical entries
![image](https://github.com/user-attachments/assets/8b334cb5-a58f-4f7f-a544-ac206f1d96f0)
(two-page document with two attachments on 1st page, and one attachment on 2nd page)

- only one file item encoded in the PDF internally, so the total file size is independent on the count of identical attachment.


Note: this result can be achieve in the Adobe Acrobat also, i.e. it's not Apache FOP result only.

Further development tasks to prevent the duplicate displaying:
- [ ] remove item with the value `Attachment tab` in the column `Location in document`
- [ ] if the file attached, but doesn't have the link in the document, then item shouldn't be removed

**Comment by Intelligent2013** [comment]
Very strange, the resulted PDF with `FileAttachment` isn't PDF/A-3 valid:
```
TestAssertion [ruleId=RuleId [specification=ISO 19005-3:2012, clause=6.3.3, testNumber=1], status=failed,
message=Every annotation (including those whose Subtype value is Widget, as used for form fields), except for the two cases listed below, shall have at least one appearance dictionary: - annotations where the value of the Rect key consists of an array where value 1 is equal to value 3 and value 2 is equal to value 4; - annotations whose Subtype value is Popup or Link,
location=Location [level=CosDocument, context=root/document[0]/pages[0](11 0 obj PDPage)/annots[0](22 0 obj PDFileAttachmentAnnot)], locationContext=null, errorMessage=null],
```
but the method `annotation.constructAppearances();` called. The objects `PDAnnotationFileAttachment` doesn't contain `/AP` values. I'll investigate it.

**Comment by Intelligent2013** [comment]
> The objects `PDAnnotationFileAttachment` doesn't contain `/AP` values.

Fixed.

**Comment by Intelligent2013** [comment]
Current issue - the annotation `FileAttachment` points to the object `6 0 R` (`/Filespec`) instead of `18 0 obj` (`/EmbeddedFile`) directly:
```
22 0 obj
<<
/Type /Annot
/Subtype /FileAttachment
/FS <<
/EF <<
/F 6 0 R
>>
/F (_test_attachments.tc7_attachments/document.presentation.pdf)
/Type /F
/AFRelationship /Data
>>
/Contents ([NO INFORMATION AVAILABLE])
/Name /Paperclip
/Rect [353.732 643.155 371.732 661.155]
/C [0 0 0]
/Border [0 0 0]
/H /I
/F 28
/AP 45 0 R
>>
endobj
```

```
6 0 obj
<<
/Type /Filespec
/F (_test_attachments.tc7_attachments/document.presentation.pdf)
/UF (_test_attachments.tc7_attachments/document.presentation.pdf)
/AFRelationship /Data
/EF <<
/F 18 0 R
>>
/Desc ([NO INFORMATION AVAILABLE])
>>
endobj
```

```
18 0 obj
<<
/Length 2209227
/Type /EmbeddedFile
/Subtype /application#2Foctet-stream
/Params <<
/CreationDate (D:20240919212649+03'00')
/ModDate (D:20240919212649+03'00')
/Size 2288366
>>
/Filter /FlateDecode
>>
stream
...
```
The Attachment tab shows all references (with 0 bytes size for annotations):
![image](https://github.com/user-attachments/assets/2882683f-7453-4456-9269-e129904d895f)
but the links aren't working.

Looks like the Apache FOP doesn't allow to retrieve the object ID from `/EF` entry. I'll investigate how to solve it.

**Comment by Intelligent2013** [comment]
> * [ ] remove item with the value `Attachment tab` in the column `Location in document`
> * [ ] if the file attached, but doesn't have the link in the document, then item shouldn't be removed

Done.

**Comment by Intelligent2013** [comment]
The example PDF (link isn't working): [test_attachments.presentation.pdf](https://github.com/user-attachments/files/17090977/test_attachments.presentation.pdf)

After opening in Adobe Acrobat - two icons - big from `/AP` entry (added for PDF-A/3 conformance), small from Acrobat:
![image](https://github.com/user-attachments/assets/73417b1a-ee1f-40af-b2f5-7d48787430e6)

After press on 'Enable editing' - only small icon:
![image](https://github.com/user-attachments/assets/822a28c4-417b-4971-87e4-f296944361ed)

Looks like I have to put the empty `/AP` entry.

**Comment by Intelligent2013** [comment]
No, both icons from `/AP` entry. I've changed one coordinate in the method `drawPaperclip`:
![image](https://github.com/user-attachments/assets/7140fc4c-ca4d-4358-8ecc-d4b4221d1868)

**Comment by petervwyatt** [comment]
@Intelligent2013 

Some issues on page 5: the file attachment annotation **/FS** entry has an incorrect **/Type** - it is currently _/F_ but should be _/Filespec_ as per Table 43 in ISO 32000-2:2020. Also the **/F** entry has a string value which includes an O/S path portion - this will cause many PDF apps issues as this used to be an attack vector (to get to `/etc/passwd`, etc!) - please make it just `(program.c)`
```
<<
	/AFRelationship 	/Data 
	/EF 	<<
		/F 	 6 0 R  
	>> 
	/F 	(_test_attachments_attachments/program.c) 
	/Type 	/F 
>>
```
Also, for PDF/A conformance, ensure that the **/EF** sub-dictionary also has a **/UF** entry that is a duplicate of the **/F** entry - in this case `/F 6 0 R`. **UF** is for Unicode but since this is a pure ASCII filename you're safe to simply duplicate.

**Comment by petervwyatt** [comment]
It would also appear that the "double paperclips" is a glitch in Adobe's implementation - others show exactly 1 paperclip but at a fixed magnification. Maybe check your flags and/or matrix???
![image](https://github.com/user-attachments/assets/4a61a63c-6803-46fc-aca1-003a653fbc51)
![image](https://github.com/user-attachments/assets/4b74dbe7-1b19-43da-8bb2-35a048425718)

**Comment by Intelligent2013** [comment]
@petervwyatt thank you for your comments! They're very useful.

> It would also appear that the "double paperclips" is a glitch in Adobe's implementation - others show exactly 1 paperclip but at a fixed magnification. Maybe check your flags and/or matrix???

Actually I have to turn off 'nozoom' flag and now the single paperclips is showing before and after 'Enable editing'.

[test_attachments.presentation.pdf](https://github.com/user-attachments/files/17101339/test_attachments.presentation.pdf)

**Comment by petervwyatt** [comment]
That new PDF looks good as far as file attachments are concerned.

For tagging purposes, you will also need to add a **StructParent** entry to the FileAttachment annot (just like you already do for Link annots) and include it in the structure tree as an **Annot** structure type.

I also had Acrobat | Print Production | Preflight | Essentials | "Check for syntax errors" report one error (and which did not show up when validating for PDF/A-3) - but it is not related to file attachments and I did _not_ confirm if it is a real error or not:
![image](https://github.com/user-attachments/assets/363a8951-31c4-4c6d-a03c-372385862672)

**Comment by Intelligent2013** [comment]
> For tagging purposes, you will also need to add a **StructParent** entry to the FileAttachment annot (just like you already do for Link annots) and include it in the structure tree as an **Annot** structure type.

@petervwyatt thank you for the notice. It was ny fault. Fixed.

[test_attachments.presentation.pdf](https://github.com/user-attachments/files/17135584/test_attachments.presentation.pdf)

![image](https://github.com/user-attachments/assets/2c62c3b3-0f57-4c4c-a596-7a4052de0b5f)

**Comment by Intelligent2013** [comment]
Done in https://github.com/metanorma/mn2pdf/releases/tag/v2.01.

**Comment by petervwyatt** [comment]
@Intelligent2013 - listing this here only because it relates to the updated PDF in the comment directly above. If you want to create a new issue please do so...

If you run it through the veraPDF Docker and select the PDF/UA-1 conformance check, it notes that the File attachment annotation is not contained inside an Annot tag: "_An annotation, excluding annotations of subtype Widget, PrinterMark or Link, shall be nested within an Annot tag_"

If you also run Adobe Acrobat preflight check for PDF/UA-1 you will get 2 failure messages that overlap with veraPDF:

* the first is for all 3 Link annotations which are missing the **Contents** key (see Table 166, ISO 32000-2:2020). If it is a URI, you can set this to the URL or do something smarter like "Visit <url>" or "Email <address>" (since you have a mailto: URI). For the ToC, set it to the heading but be sure to include a single space between heading numbers and title. 

* the second is for the non-visual "table" on the ISO front cover which is irregular and incorrectly tagged from a PDF/UA perspective. If you use other validator tools such as [PAC](https://pac.pdf-accessibility.org/en) or https://pdf4wcag.com/ they will explain a bit more about what is wrong with this table (PAC will also show you).

**Comment by Intelligent2013** [comment]
Fixed.
